standard/expressions.md

```diff
@@ -6201,6 +6368,49 @@ When a property or indexer declared in a *struct_type* is the target of an assig
 >
 > *end example*
 
+### §ref-assignment-new-clause Ref assignment
+
+The `= ref` operator is known as the *ref assignment* operator.
+
+The left operand shall be an expression that binds to a reference variable (§ref-span-safety), a reference parameter (other than `this`), an output parameter, or and input parameter. The right operand shall be an expression that yields a *variable_reference* designating a value of the same type as the left operand.
+
+It is a compile time error if the ref-safe-context of the left operand is wider than the ref-safe-context of the left operand (§ref-safe-contexts).
+
+The right operand shall be definitely assigned at the point of the ref assignment.
+
+When the left operand binds to an `out` parameter, it is an error if that `out` parameter has not been definitely assigned at the beginning of the ref assignment operator.
+
+If the left operand is a writeable ref (i.e., it designates anything other than a `ref readonly` local or  `in` parameter), then the right operand shall be a writeable *variable_reference*. If the right operand variable is writeable, the left operand may be declared `ref` or `ref readonly`.
+
+The operation makes the left operand an alias of the right operand variable. The alias may be made read-only even if the right operand variable is writeable.
+
+The ref assignment operator yields a *variable_reference* of the assigned type. It is writeable if the left operand is writeable.
+
+The ref assignment operator shall not read the storage location referenced by the right hand operand.
+
+> *Example*: Here are some examples of using `= ref`:
+>
+> ```csharp
+> public static int M1() { … }
+> public static ref int M2() { … }
+> public static ref uint M2u() { … }
+> public static ref readonly int M3() { … }
+> 
+> int v = 42;
+> ref int r1 = ref v; // OK, r1 refers to v, which has value 42
+> r1 = ref M1();      // Error; M1 returns a value, not a reference
+> r1 = ref M2();      // OK; makes an alias
+> r1 = ref M2u();     // Error; lhs and rhs have different types
+> r1 = ref M3();    // error; M3 returns a ref readonly, which r1 cannot honor
+> ref readonly r2 = ref v; // OK; make readonly alias to ref
+> r2 = ref M2();      // OK; makes an alias, adding read-only protection
+> r2 = ref M3();      // OK; makes an alias and honors the read-only
+> r2 = ref (r1 = ref M2());  // OK; r1 is an alias to a writable variable,
+>               // r2 is an alias (with read-only access) to the same variable
+> ```
+>
+> *end example*
+
```

Comment:
I had an action item to add a note here. I don't think we *have* to do before we ship C# 7, but it *could* be something like:

```text
*Note*: When reading code using an `= ref` operator, it can be tempting to read
the `ref` part as being part of the operand. This is particularly confusing when
the operand is a conditional `?:` expression. For example, when reading
`ref int a = ref b ? ref x : ref y;` it's important to read this as `= ref` being the operator,
and `b ? ref x : ref y` being the right operand. Importantly, the expression
`ref b` is *not* part of that statement, even though it might appear so at first glance.
```

I'm not convinced that's good enough to include yet, but I'd welcome feedback about whether it's close enough to be included with a bit of mob-wordsmithing.

Reply:
I like it. I think it's clarified with the addition of parentheses in the example, after explaining how it's parsed:

```csharp
ref int a = ref (b ? ref x : ref y);
```

Reply:
Yes! That definitely helps a lot.

Reply:
Suggestion committed.